specs/ec/ec.tla

```diff
@@ -10,118 +17,158 @@ Range(f) ==
 
 (* --algorithm ExpectedConsensus
 variables
-    NumMiners=5;
-    miners = 1..NumMiners;
-    MaxRounds=10;
-    MaxLeaders = 3;
-    rounds = 1..MaxRounds;
-    miner_rounds = [miner \in miners |-> 0];
-    leaders = <<{}>>;
+    miner_rounds = [miner \in miners |-> 2];
+    leaders = <<{<<CHOOSE miner \in miners: TRUE, {}>>}, {}>>; \* Let someone win the first round.
     
 define
     FairMining ==
         \A miner \in miners: miner_rounds[miner] - Min(Range(miner_rounds)) <= 1
 
     BoundedLeaders == 
         ~\E leader_group \in Range(leaders): Cardinality(leader_group) > MaxLeaders
-        
+    ModelBoundedLeaders == 
+        ~\E leader_group \in Range(leaders): Cardinality(leader_group) > MaxLeaders - 1
+    
+    \* This invariant is used to force an error trace after MaxRounds.
     NoProgress ==
-        miner_rounds[CHOOSE m \in miners: TRUE] < 11
-        
-    Scratch(miner) ==
+        /\ Len(leaders) < MaxRounds 
+        /\ miner_rounds \in [miners -> 0..MaxRounds]
+
+    Scratch(miner, tipset) ==
         \* all have equal chance of winning
-        LET e == RandomElement(miners) IN
-            e = miner
+        RandomElement(miners) = miner
 
 end define;
 
-process miner \in miners
+fair process miner \in miners
 begin
     Start:
-        await miner_rounds[self] <= Min(Range(miner_rounds));
-        if Scratch(self) then
-            leaders := <<Head(leaders) \union {self}>> \o Tail(leaders);
-        else 
-            skip;
-        end if;
-        
-    Tick:
+        await /\ miner_rounds[self] = Min(Range(miner_rounds))
+              /\ Len(leaders) >= miner_rounds[self] - 1;      
+        if miner_rounds[self] >= MaxRounds then goto Done end if;
+    Elect:
+        with current_leaders = leaders[miner_rounds[self] - 1] do 
+            if current_leaders = {} then
+                leaders[Len(leaders)] := leaders[Len(leaders)] \union {<<self, {}>>};
+            
+                skip; \* TODO: Handle null mining.
+            else
+                with tipset \in SUBSET {x[1] : x \in current_leaders }  do
+                    if USE_RANDOM then
```

Comment:
You're modeling both with RANDOM and BOOLEAN?

Reply:
It’s configurable to allow swapping between simulation mode and an exhaustive model.